### PR TITLE
BinXor implementation

### PR DESCRIPTION
**Pull request by gostkin**

### CODE REVIEW

File: src/main/scala/sigmastate/eval/Evaluation.scala [hunk 1]
```diff
@@ -404,6 +404,9 @@ trait Evaluation extends RuntimeCosting { IR =>
               out(th())
             } else
               out(false)
+          case ApplyBinOpLazy(op, In(x: Boolean), In(y)) if op == BinaryXorOp =>
+            val th = y.asInstanceOf[() => Boolean]
+            out(th() ^ x)
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/Evaluation.scala]: @gostkin I'm not sure it should be lazy. Let's check how it is done in Scala, and do the same.

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 2]
```diff
@@ -1692,12 +1692,12 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
-//      case BinXor(l, r) =>
-//        val lC = evalNode(ctx, env, l)
-//        val rC = RCostedThunk(Thunk(evalNode(ctx, env, r)), 0)
-//        val v = sigmaDslBuilder.binXor(lC.value, rC.value)
-//        val c = lC.cost + rC.cost + costOf(node)
-//        withDefaultSize(v, c)
+      case BinXor(l, r) =>
+        val lC = evalNode(ctx, env, l)
+        val rC = RCostedThunk(Thunk(evalNode(ctx, env, r)), 0)
+        val v = BinaryXorOp.applyLazy(lC.value, rC.value)
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: See comment about about lazyness

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 3]
```diff
@@ -1692,12 +1692,12 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
+        val c = opCost(Seq(lC.cost, rC.cost), costOf(node))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: Please use opCost(v, Seq(lC.cost, rC.cost), costOf(node)) version
Please see this comment https://github.com/ScorexFoundation/sigmastate-interpreter/pull/477#discussion_r278424120

File: src/main/scala/sigmastate/lang/SigmaTyper.scala [hunk 4]
```diff
@@ -316,14 +316,22 @@ class SigmaTyper(val builder: SigmaBuilder, predefFuncRegistry: PredefinedFuncRe
...
+              }
               res
             case SSigmaProp =>
-              val (a,b) = (newObj.asSigmaProp, r.asSigmaProp)
-              val res = if (m == "||") mkSigmaOr(Seq(a,b)) else mkSigmaAnd(Seq(a,b))
+              val (a, b) = (newObj.asSigmaProp, r.asSigmaProp)
+              val res = m match {
+                case "||" => mkSigmaOr(Seq(a, b))
+                case "&&" => mkSigmaAnd(Seq(a, b))
+                case "^" => throw new NotImplementedError(s"Not implemented") // TODO: mkSigmaXor(Seq(a, b))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/lang/SigmaTyper.scala]: let's make more explicit accurate message that operation is not defined with all context information

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala (hunk 3 above)
Comment (gostkin) [reply src/main/scala/sigmastate/eval/RuntimeCosting.scala]: This change leads to fail
```
scalan.TypeDescs$$anon$7 cannot be cast to special.collection.impl.CostsDefs$Costed$CostedElem
java.lang.ClassCastException: scalan.TypeDescs$$anon$7 cannot be cast to special.collection.impl.CostsDefs$Costed$CostedElem
```

**Pull request merged**